js-packages/@prestojs/rest/src/CursorPaginator.ts

```diff
@@ -0,0 +1,116 @@
+import { EndpointExecuteOptions } from './Endpoint';
+import Paginator, { PaginationStateChangeListener } from './Paginator';
+
+export type CursorPaginationState = {
+    pageSize?: number;
+    cursor?: string;
+};
+
+export default class CursorPaginator extends Paginator {
+    nextCursor: string | null = null;
+    previousCursor: string | null = null;
+
+    get cursor(): string | null | undefined {
+        return this.currentState.cursor;
+    }
+
+    get pageSize(): number | null | undefined {
+        return this.currentState.pageSize;
+    }
+
+    constructor(
+        initialState: CursorPaginationState = {},
+        onChange?: PaginationStateChangeListener
+    ) {
+        super(initialState || {}, onChange);
+    }
+
+    setPageSize(pageSize: number): void {
+        if (pageSize < 1) {
+            throw new Error(`Invalid pageSize ${pageSize} - should be >= 1`);
+        }
+        this.setState(currentState => {
+            // If page size changes any next/previous cursors are no invalid
+            if (currentState.pageSize !== pageSize) {
+                this.nextCursor = null;
+                this.previousCursor = null;
+            }
+            // When page size changes we need to reset the cursor
+            if (currentState.pageSize !== pageSize && currentState.cursor) {
+                const nextState: CursorPaginationState = { ...currentState, pageSize };
+                delete nextState.cursor;
+                return nextState;
+            }
+            return { ...currentState, pageSize };
+        });
+    }
+
+    first(): void {
+        // first page is equivalent to not specifying cursor
+        this.setState(currentState => {
+            const nextState: CursorPaginationState = { ...currentState };
+            delete nextState.cursor;
+            return nextState;
+        });
+    }
+
+    next(): void {
+        if (this.nextCursor == null) {
+            throw new Error('Cannot go to next page as next cursor is not known');
```

Comment:
Can you expand on this more because I don't really follow? 

Are you talking about returning a link to the _api_ or to a link to the frontend that then calls the api? I see no reason you can't do the latter (there's an example of how you do this although it relies on a hook that you'd currently have to write - search for `useUrlState`)

Reply:
(I think this is explained in your other comment that I have now responded to)